### PR TITLE
ddl: support bdr bwlist

### PR DESCRIPTION
**Pull request by okJiang**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/48519

Problem Summary:

### What is changed and how it works?

- restricted DDL not from CDC by BDR role
- Categorized all DDL to safeDDL/unsafeDDL/unmanagedDDL
  1) add `BDRActionMap`(BDRType => []Action), 
  2) add `ActionBDRMap`(Action => BDRType)

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)

image

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/parser/model/ddl.go [hunk 1]
```diff
@@ -180,9 +181,97 @@ var actionMap = map[ActionType]string{
...
+	// SafeDDL means the DDL can be executed by user when cluster is Primary.
+	SafeDDL DDLBDRType = "safe DDL"
+	// UnmanagementDDL means the DDL can't be synced by CDC.
+	UnmanagementDDL DDLBDRType = "unmanagement DDL"
+	// UnknownDDL means the DDL is unknown.
+	UnknownDDL DDLBDRType = "unknown DDL"
+)
+
+// ActionBDRMap is the map of DDL ActionType to DDLBDRType.
+var ActionBDRMap = map[ActionType]DDLBDRType{
```
Comment (asddongmen) [inline pkg/parser/model/ddl.go:199]: Do you think it's better to group the them by their DDLBDRType?

Comment (okJiang) [reply pkg/parser/model/ddl.go:199]: I think your concern is that `group the them by their DDLBDRType` can more visually see its classification.

But that `group by ActionType` is more easily used in code.

So, I can add another Map to show its classification, and add a new ut to keep them consistent.

File: pkg/parser/model/ddl.go [hunk 2]
```diff
@@ -180,9 +181,177 @@ var actionMap = map[ActionType]string{
...
```
Comment (Benjamin2037) [inline pkg/parser/model/ddl.go:199]: Why do we need ActionBDRMap and BDRActionMap？

Comment (okJiang) [reply pkg/parser/model/ddl.go:199]: see https://github.com/pingcap/tidb/pull/48776#discussion_r1410277095

File: pkg/parser/model/ddl.go [hunk 3]
```diff
@@ -180,9 +182,177 @@ var actionMap = map[ActionType]string{
...
+		ActionMultiSchemaChange,
+		ActionFlashbackCluster,
+		ActionRecoverSchema,
+		ActionReorganizePartition,
+		ActionAlterTablePartitioning,
+		ActionRemovePartitioning,
+	},
+	UnmanagementDDL: {
+		ActionCreateView,
+		ActionDropView,
```
Comment (okJiang) [inline pkg/parser/model/ddl.go:340]: Should they be classified into safe DDL? @asddongmen 

image

File: pkg/parser/model/ddl.go (hunk 1 above)
Comment (tangenta) [reply pkg/parser/model/ddl.go:200]: Maybe we can initialize `ActionBDRMap` in `init()` like https://github.com/pingcap/tidb/blob/7ef4b4bb69a2e0d90ee19c6b23c7be8ec8ce2e0d/pkg/expression/builtin_convert_charset.go#L300

Thus, the UT is not necessary.

File: pkg/session/bootstrap.go [hunk 4]
```diff
@@ -1043,11 +1042,15 @@ const (
 	// version 181
 	//   add a new `type` column on mysql.bind_info, which is for universal binding #48875.
 	version181 = 181
+
+	// vresion 182
```
Comment (tangenta) [inline pkg/session/bootstrap.go:1046]:
```suggestion
	// version 182
```

File: pkg/parser/model/ddl.go (hunk 3 above)
Comment (okJiang) [reply pkg/parser/model/ddl.go:269]: yes, it is safe DDL

File: pkg/parser/model/ddl.go [hunk 5]
```diff
@@ -448,6 +547,12 @@ type Job struct {
 
 	// TraceInfo indicates the information for SQL tracing
 	TraceInfo *TraceInfo `json:"trace_info"`
+
+	// BDRRole indicates the role of BDR cluste when executing this DDL.
+	BDRRole string `json:"bdr_role"`
+
+	// CDCWriteSource indicates the source of CDC write.
+	CDCWriteSource uint64 `json:"cdc_write_source"`
```
Comment (tangenta) [inline pkg/parser/model/ddl.go:555]: Maybe add "omitempty" to these two json tags?

Comment (okJiang) [reply pkg/parser/model/ddl.go:555]: What is your concern?

Comment (tangenta) [reply pkg/parser/model/ddl.go:555]: OK. This is only helpful when BDRRole is zero in most cases.

File: pkg/parser/model/ddl.go [hunk 6]
```diff
@@ -448,6 +547,12 @@ type Job struct {
 
 	// TraceInfo indicates the information for SQL tracing
 	TraceInfo *TraceInfo `json:"trace_info"`
+
+	// BDRRole indicates the role of BDR cluste when executing this DDL.
```
Comment (tangenta) [inline pkg/parser/model/ddl.go:551]:
```suggestion
	// BDRRole indicates the role of BDR cluster when executing this DDL.
```

**Review by tangenta — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
schema change is good for now but need to follow up with version upgrade/rollback process.

**Comment by yudongusa** [comment]
> schema change is good for now but need to follow up with version upgrade/rollback process.

@Benjamin2037

**Comment by Benjamin2037** [comment]
> > schema change is good for now but need to follow up with version upgrade/rollback process.
> 
> @Benjamin2037

Add column for the system table will not cause the compatible problem of the rollback TiDB version， this is changing the DDL-related system table, and we will auto-pause users' DDL tasks. it will not cause problems.

**Pull request merged**